core/tests/protractor_utils/general.js

```diff
@@ -35,6 +35,13 @@ var waitForSystem = function() {
   browser.sleep(waitTime);
 };
 
+var waitForLoadingMessage = function() {
```

Comment:
Navigating to a new page without Angular completely loading will cause tpload error

Reply:
Might be worth making this an inline comment.

Reply:
(Meta-comment: Please respond to all open comments -- see step 5 [here](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change). I can't tell whether this is resolved or not.)

Reply:
That was weird. I thought I already reached out to all comments.... my bad!

So here is the added comment:
```javascript
  // Consider adding this method after each browser.get() call going to an
  // Angular page destination. Completely wait for page to load to
  // avoid XMLHTTPReq error on page refresh:
  // https://github.com/angular/angular.js/issues/14219#issuecomment-251605766
  // and browser.waitForAngular()'s flakiness
  // https://github.com/angular/protractor/issues/2954.
```
So the first issue is the XMLHTTPRequest error. You can reproduce by chaining 2 browser.get() calls together in protractor, like so
```javascript
browser.get(general.SERVER_URL_PREFIX + '/library');
browser.get(general.SERVER_URL_PREFIX + '/learner_dashboard');
```
Waiting on `$rootscope.loadingMessage`  is key here. It is how the current page signals the necessary components are loaded from the backend completely and is ready.  

Also the 2nd issue is waitForAngular 's interaction with Oppia internal components. The added full stop messed up the [link](https://github.com/angular/protractor/issues/2954). 

From my point of view, waitForAngular can potentially be misused due to non-deterministic expectation. It can be considered "flaky" because sometimes waitForAngular can be long enough or sometimes not at all for Oppia. For Protractor to know what it is waiting for, one needs to be specify which element is supposed to be waited for, which is the "loadingMessage" element to be hidden.